### PR TITLE
Use BigNumbers

### PR DESCRIPTION
**Pull request by emschwartz**
This switches the module to read and write `bignumber.js` BigNumbers instead of JS numbers or the `[high, low]` form for UInt64s.

Most modules that use this library use BigNumber math anyway, and there are a number that needlessly include the [`long`](https://github.com/dcodeIO/long.js) module just to convert BigNumbers to the `[high, low]` format this library uses. This library also tried reading VarInts into JS numbers and would throw an error if the number was too large, which is very unhelpful if a protocol allows VarInts to be used.

Originally, I was considering trying to make this change non-breaking and either add `returnBigNumbers` as an optional flag or making the library return BigNumbers only for values larger than the maximum safe JS integer. However, after discussing with @sharafian we agreed that it would be less surprising if we just made all of the read functions return BigNumbers (even for UInt8's, UInt16's, and so on).

Resolves https://github.com/interledgerjs/oer-utils/issues/10

### CODE REVIEW

File: src/lib/reader.ts [hunk 1]
```diff
@@ -91,39 +87,34 @@ class Reader {
...
     }
-    if (length > Reader.MAX_INT_BYTES) {
-      throw new Error('Tried to read too large integer (requested: ' +
-        length + ', max: ' + Reader.MAX_INT_BYTES + ')')
-    }
-    this.ensureAvailable(length)
-    const value = this.buffer.readUIntBE(this.cursor, length)
-    return value
+
+    return new BigNumber(this.peek(length).toString('hex'), 16)
```
Comment (sharafian) [inline src/lib/reader.ts]: This seems slightly janky; couldn't we compute the number manually? Something like:

```js
return this.peek(length).reduce((sum, byte) =>  sum.times(256).plus(byte), new BigNumber(0))
```

File: src/lib/reader.ts [hunk 2]
```diff
@@ -133,87 +124,50 @@ class Reader {
...
       throw new Error('Tried to read integer with negative length (provided: ' +
         length + ')')
+    } else if (length > 8) {
+      throw new Error('Ints longer than 8 bytes must be encoded as VarInts')
     }
-    if (length > Reader.MAX_INT_BYTES) {
-      throw new Error('Tried to read too large integer (requested: ' +
-        length + ', max: ' + Reader.MAX_INT_BYTES + ')')
+
+    const value = new BigNumber(this.peek(length).toString('hex'), 16)
```
Comment (sharafian) [inline src/lib/reader.ts]: I think we should calculate this instead of using toHex

File: test/writer.spec.ts [hunk 3]
```diff
@@ -359,13 +400,20 @@ describe('Writer', function () {
...
-      )
+      writer.writeVarInt(new BigNumber('0100000000000000', 16))
+
+      assert.equal(writer.getBuffer().toString('hex'), '080100000000000000')
+    })
+
+    it('should write a large (absolute value) negative integer', function () {
+      const writer = new Writer()
+
+      writer.writeVarInt(new BigNumber('010000000000000000000000000001', 16))
```
Comment (sharafian) [inline test/writer.spec.ts]: this doesn't look negative, am I missing something?

File: README.md [hunk 4]
```diff
@@ -25,18 +37,19 @@ const Reader = require('oer-utils/reader')
 const reader = Reader.from(new Buffer('1234', 'hex'))
 
 const v1 = reader.readUInt8()
-const v2 = reader.readUInt8()
+const v2 = reader.readUInt8BigNum()
```
Comment (sentientwaffle) [inline README.md]: What do you think of `readBigUInt8()` for the naming pattern? (slightly shorter, and a little easier to read)

Comment (emschwartz) [reply README.md]: I think I prefer it at the end. `BigUInt8` sounds like the UInt8 is big, but that's a bit of a contradiction. It's more like `readUInt8AsBigNumber`, but that seems a little too verbose

**Comment by sublimator** [comment]
wah, only strings and bignums, even for <= 32 bit numbers ? 

hrmmmm

File: src/lib/util.ts [hunk 5]
```diff
@@ -1,6 +1,31 @@
...
+
+export function bufferToBigNumber (buffer: Buffer): BigNumber {
+  return buffer.reduce(
+    (sum, value) => sum.times(256).plus(value),
+    new BigNumber(0)
+  )
+}
+
+export function bigNumberToBuffer (value: BigNumber, length?: number): Buffer {
+  const lengthOfValue = (length !== undefined ? length : Math.ceil(value.toString(2).length / 8))
```
Comment (sublimator) [inline src/lib/util.ts]: value.toString(2) and not value.toString(16) ??

File: src/lib/util.ts [hunk 6]
```diff
@@ -1,6 +1,31 @@
...
+    return typeof value !== 'object'
+      && typeof value !== 'function'
+      && isFinite(value)
+      && Math.floor(value) === value
+  }
+}
+
+export function bufferToBigNumber (buffer: Buffer): BigNumber {
+  return buffer.reduce(
+    (sum, value) => sum.times(256).plus(value),
```
Comment (sublimator) [inline src/lib/util.ts]: new BigNumber(buffer.toString('hex'), 16) doesn't work?

File: src/lib/util.ts (hunk 5 above)
Comment (emschwartz) [reply src/lib/util.ts]: Oops, right

**Comment by emschwartz** [comment]
> wah, only strings and bignums, even for <= 32 bit numbers ?

Yeah, I was on the fence on that one. It seemed more consistent, but we could have also made 8, 16, and 32 bit integers use JS numbers

File: src/lib/util.ts (hunk 6 above)
Comment (emschwartz) [reply src/lib/util.ts]: It does. I had that originally and @sharafian thought that was more janky

**Comment by sublimator** [comment]
I have been using this lib for a BLE peripheral library built on top of bleno, for encoding the messages :)

File: src/lib/util.ts (hunk 6 above)
Comment (sublimator) [reply src/lib/util.ts]: Ugh, lol, why would you listen to a guy who doesn't even drink his vodka and tomato juices :)

**Pull request merged**